c7n/resources/ssm.py

```diff
@@ -597,3 +598,117 @@ def register_resource(cls, registry, resource_class):
+
+
+@resources.register('ssm-document')
+class SSMDocument(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'ssm'
+        enum_spec = ('list_documents', 'DocumentIdentifiers', None)
+        name = 'Name'
+        date = 'RegistrationDate'
+        arn_type = 'Document'
+
+    permissions = ('ssm:DescribeDocumentPermission', 'ssm:ListDocuments', )
+
+
+@SSMDocument.filter_registry.register('cross-account')
+class SSMDocumentCrossAccount(CrossAccountAccessFilter):
+    """Filter SSM documents which have cross account permissions
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: ssm-cross-account
+                resource: ssm-document
+                filters:
+                  - type: cross-account
+                    whitelist: [xxxxxxxxxxxx]
+    """
+
+    permissions = ('ssm:DescribeDocumentPermission',)
+
+    def process(self, resources, event=None):
+        self.accounts = self.get_accounts()
+        results = []
+        client = local_session(self.manager.session_factory).client('ssm')
+        with self.executor_factory(max_workers=3) as w:
+            futures = []
+            for resource_set in chunks(resources, 10):
+                futures.append(w.submit(
+                    self.process_resource_set, client, resource_set))
+            for f in as_completed(futures):
+                if f.exception():
+                    self.log.error(
+                        "Exception checking cross account access \n %s" % (
+                            f.exception()))
+                    continue
+                results.extend(f.result())
+        return results
+
+    def process_resource_set(self, client, resource_set):
+        results = []
+        for r in resource_set:
+            if r['Owner'] == 'Amazon':
+                continue
+            attrs = self.manager.retry(
+                client.describe_document_permission,
+                Name=r['Name'],
+                PermissionType='Share')['AccountSharingInfoList']
+            shared_accounts = {
+                g.get('AccountId') for g in attrs}
+            delta_accounts = shared_accounts.difference(self.accounts)
+            if delta_accounts:
+                r['c7n:CrossAccountViolations'] = list(delta_accounts)
+                results.append(r)
+        return results
+
+
+@SSMDocument.action_registry.register('remove-sharing')
+class RemoveSharingSSMDocument(Action):
+    """Remove external accounts from SSM document permissions
+    """
+
+    schema = type_schema('remove-sharing')
+    permissions = ('ssm:ModifyDocumentPermission',)
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('ssm')
+        for r in resources:
+            client.modify_document_permission(
+                Name=r['Name'],
+                PermissionType='Share',
+                AccountIdsToRemove=r['c7n:CrossAccountViolations']
```

Comment:
would also be useful to support the ability to pass in specific account ids to remove